packages/contracts-bedrock/contracts/universal/ProxyAdmin.sol

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import { Proxy } from "./Proxy.sol";
+import { Owned } from "@rari-capital/solmate/src/auth/Owned.sol";
+import { Lib_AddressManager } from "../legacy/Lib_AddressManager.sol";
+import { L1ChugSplashProxy } from "../legacy/L1ChugSplashProxy.sol";
+import { Bytes32AddressLib } from "@rari-capital/solmate/src/utils/Bytes32AddressLib.sol";
+
+
+/**
+ * @title ProxyAdmin
+ * @dev This is an auxiliary contract meant to be assigned as the admin of a
+        Proxy, based on the OpenZeppelin implementation.
+ */
+contract ProxyAdmin is Owned {
+
+   enum ProxyType {
+       OpenZeppelin,
+       Chugsplash,
+       ResolvedDelegate
+   }
+
+   mapping(address => ProxyType) internal _proxyType;
+   mapping(address => string) internal _proxyName;
+
+   Lib_AddressManager addressManager;
+
+    /**
+     * @notice A legacy upgrading indicator used by the old Chugsplash Proxy
+     * @custom:legacy
+     */
+    bool internal upgrading = false;
+
+    /**
+     * @notice Set the owner of the ProxyAdmin via constructor argument.
+     */
+
+    constructor(address owner) Owned(owner) {}
+
+    /**
+     * @notice
+     *
+     * @param _address   The address of the proxy
+     * @param _type The type of the proxy
+     */
+    function setProxyType(address _address, ProxyType _type) public onlyOwner {
+        _proxyType[_address] = _type;
+    }
+
+    function getProxyType(address _address) public view returns (ProxyType) {
+        return _proxyType[_address];
+    }
+
+    function setProxyName(address _address, string memory _name) public onlyOwner {
+        _proxyName[_address] = _name;
+    }
+
+    function getProxyName(address _address) public view returns (string memory) {
+        return _proxyName[_address];
+    }
+
+    function setAddressManager(address _address) external onlyOwner {
+        addressManager = Lib_AddressManager(_address);
+    }
+
+    /**
+     * @notice
+     *
+     * @param _upgrading
+     */
+    function setIsUpgrading(bool _upgrading) external onlyOwner {
+        upgrading = _upgrading;
+    }
+
+    /**
+     * @dev Returns the current implementation of `proxy`.
+     *      This contract must be the admin of `proxy`.
+     *
+     * @param proxy The Proxy to return the implementation of.
+     * @return The address of the implementation
+     */
+    function getProxyImplementation(Proxy proxy) external view returns (address) {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        // We need to manually run the static call since the getter cannot be flagged as view
+        if (proxyType == ProxyType.OpenZeppelin) {
+            // bytes4(keccak256("implementation()")) == 0x5c60da1b
+            (bool success, bytes memory returndata) = address(proxy).staticcall(hex"5c60da1b");
```

Comment:
Might be better to just wrap the proxy in the interface to avoid needing to manually encode stuff

Reply:
+ Same comment where used elsewhere

Reply:
This will not work because of the possibility of the delegatecall that may modify state